Software/real_time/ROS_RoboBuggy/src/robobuggy/src/imu/IMU_Broadcaster.cpp

```diff
@@ -1,30 +1,126 @@
 #include "imu/IMU_Broadcaster.h"
 
-float IMU_Broadcaster::get_spoofed_x_accel()
-{
-    return 1.5;
-}
-float IMU_Broadcaster::get_spoofed_y_accel()
-{
-    return 2.5;
-}
-float IMU_Broadcaster::get_spoofed_z_accel()
+IMU_Broadcaster::IMU_Broadcaster()
 {
-    return 1.5;
+    imu_pub = nh.advertise<robobuggy::IMU>("IMU", 1000);
 }
 
-IMU_Broadcaster::IMU_Broadcaster()
+int IMU_Broadcaster::init_IMU()
 {
-    imu_pub = nh.advertise<robobuggy::IMU>("IMU", 1000);
+    int err;
+    err = freespace_init();
+    if (err != FREESPACE_SUCCESS) {
+        ROS_ERROR_STREAM("libfreespace initialization error\n");
+        return err;
+    }
+
+    err = freespace_getDeviceList(&device, 1, &numIds);
+    if (numIds == 0) {
+        //@TODO: Keep trying to locate IMU
+        ROS_INFO_STREAM("Didn't find any devices\n");
+    }
+
+    err = freespace_openDevice(device);
+    if (err != FREESPACE_SUCCESS) {
+        ROS_ERROR_STREAM("Error opening device\n");
+        return err;
+    }
+
+    err = freespace_flush(device);
+    if (err != FREESPACE_SUCCESS) {
+        ROS_ERROR_STREAM("Error flushing device\n");
+        return err;
+    }
+
+    // Configure device for motion outputs
+    message.messageType = FREESPACE_MESSAGE_DATAMODECONTROLV2REQUEST;
+    message.dataModeControlV2Request.packetSelect = 8;  // MotionEngine Outout
+    message.dataModeControlV2Request.mode = 0;          // Set full motion
+    message.dataModeControlV2Request.formatSelect = 0;  // MEOut format 0
+    message.dataModeControlV2Request.ff0 = 1;           // Pointer fields
+	message.dataModeControlV2Request.ff1 = 1;           // Linear Acceleration
+	message.dataModeControlV2Request.ff2 = 1;           // Linear Acce no gravity
+    message.dataModeControlV2Request.ff3 = 1;           // Angular velocity fields
+	message.dataModeControlV2Request.ff4 = 1;           // Magnetometer
+	message.dataModeControlV2Request.ff6 = 1;           // Angular position
+
+    err = freespace_sendMessage(device, &message);
+    if( err != FREESPACE_SUCCESS) {
+        ROS_ERROR_STREAM("Could not send message\n");
+        return err;
+    }
+
+    return 0;
 }
 
-void IMU_Broadcaster::publish_new_spoofed_message()
+void IMU_Broadcaster::publish_IMU_message()
 {
     robobuggy::IMU msg;
+    int err;
+    
+    err = freespace_readMessage(device, &message, 100);
+
+    if ((err == FREESPACE_ERROR_TIMEOUT) || (err == FREESPACE_ERROR_INTERRUPTED)) {
+        return;
+    }
+    if (err != FREESPACE_SUCCESS) {
+        //@TODO: This doesn't actually terminate the loop
+        ROS_ERROR_STREAM("Error reading message");
+        ROS_ERROR_STREAM(err);
+        return;
+    }
+
+    if (message.messageType == FREESPACE_MESSAGE_MOTIONENGINEOUTPUT) {
+        struct MultiAxisSensor reading;
+
+        // Read acceleration data
+        err = freespace_util_getAcceleration(&message.motionEngineOutput, &reading);
+        if (err == 0) {
```

Comment:
We should print out to error if this fails, since that means that the connection is dropping messages